mantidproject/mantid
# Reflectometry GUI Restructuring 14: Catalog login

***PR ~~#25851~~ should be merged first.*** In the meantime this can be reviewed by comparing against [that branch](https://github.com/gemmaguest/mantid/compare/23027_refl_gui_update_output_values...23027_refl_gui_catalog_login)

**Description of work.**

Ongoing work to integrate the `JobTreeView` into the `ISIS Reflectometry` GUI. This PR is being merged to a feature branch because not all functionality is implemented yet.

This PR adds functionality to show the catalog login dialog if the user attempts a catalog search in the GUI while not logged in.
- Add `IPythonRunner` to provide an interface to the `runPythonAlgorithm` method in the main window view. This is also now used by the plotter to avoid having a direct pointer to the main window view.
- Pass the python runner to the `CatalogSearcher` (owned by the runs presenter).
- Improve the `CatalogSearcher` to handle login and asynchronous operation. Use this instead of the direct python calls that were previously done from the runs presenter.
- The `CatalogSearcher` uses the runs view to execute the algorithm, receives notification when it is complete, updates the results in the view, and then notifies the runs presenter.
- Tidying:
  - remove an unused `runPythonAlgorithm` function from the main window presenter;
  - remove calls to the runs view to show message boxes (use the message handler instead);
  - remove the `Autoreduction` class which is no longer required due to simplified logic in the runs presenter.

**To test:**

This should be tested by someone at ISIS.

- Ensure you are not logged in to the catalog
- Open the `ISIS Reflectometry` interface
- Select instrument `INTER` and enter investigation id `1120015`
- Click `Search`. You should be prompted to log in to the catalog.
- If you cancel the login or enter invalid credentials, the search will not proceed and you will get an error message.
- If you enter correct credentials, the search should complete and show the list of search results in the table below.
![image](https://user-images.githubusercontent.com/12895056/59444038-0f04fd80-8df5-11e9-9a2f-541ae6b2fd96.png)
- Log out of the catalog from the main menu -> catalog menu.
- Repeat but this time click `Autoprocess`. The login behaviour is the same except this time it also starts processing the runs.
![image](https://user-images.githubusercontent.com/12895056/59444084-22b06400-8df5-11e9-96cb-6342d86bb00e.png)

Refs #23027

*This does not require release notes* because **it reimplements existing functionality from the existing GUI**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
